DOCS 428: document the frame system editor (#1127)
viamrobotics/docs: docs/services/frame-system/_index.md

Any robot configured in Viam comes with the Frame System service: an internally managed and mostly static system for storing the "reference frame" of each component of a robot within a coordinate system configured by the user.

![Visualization of a wheeled base configured with motors and a mounted camera in the Frame System tab of the Viam app UI](img/frame_system_wheeled_base.png)

> 

Reviewer: On line 21 - can you add where you click add frame? on the component panel I think?

Author: Yep!

For more information about determining the appropriate values for these parameters, see these two examples:

- [A Reference Frame:](/services/frame-system/frame-config) A component attached to a static surface
- [Nested Reference Frames:](/services/frame-system/nested-frame-config) A component attached to another, dynamic, component

### Visualize the Frame System

You can visualize how your robot is oriented in the Frame System in [the Viam app](https://app.viam.com).
Navigate to the **config** tab on your robot's page, select **mode** as **Builder**, and click on **Frame System**.

> A 3D visualization of the spatial configuration of the reference frames of components configured on your robot is shown:

Reviewer: The Viam app shows you...

Author: Done

![Default frame system configuration grid visualization for a single component, shown in the Frame System Editor](img/frame_system_basic.png)

This tab provides a simple interface for simultaneously viewing and editing the position, orientation, and geometries of a robot's components in the Frame System.

> For example:

Reviewer: Unsure why all of this is in bullet points - it's all one example

> - You have configured a robot with a [`jetson` board](/components/board), wired to a [`webcam` camera](/components/camera/webcam) and a [`wheeled` base](/components/base/wheeled) with two [motors](/components/motor) driving its wheels.

Reviewer: ```suggestion
- Consider a robots configured with a [`jetson` board](/components/board), wired to a [`webcam` camera](/components/camera/webcam) and a [`wheeled` base](/components/base/wheeled) with two [motors](/components/motor) driving its wheels.
```

- You go to the **config** tab of your robot and click on the **Frame System** sub-tab.

> You have not yet added any reference frame configuration to these components, so they are all shown to be located on the default `world` origin point as follows:

Reviewer: here I'd suggest reframing this as "If you go to ..."

- You measure the distance from your board to the camera mounted overhead and find it to be 50 millimeters.
You add this value to `"Z"` in your camera's reference frame `Translation` attribute and the Frame System readjusts to show the camera's translation:

  ![Camera translated 50 mm overhead shown in the Frame System Editor](img/demo_camera_edited_1.png)

- However, the camera doesn't yet display as oriented towards the base.
You adjust the [orientation vector](/internals/orientation-vector) 0.5 degrees in `"OX"` in your camera's reference frame `Orientation` attribute and the Frame System readjusts to show the camera's orientation:

  ![Camera oriented .5 degrees OX shown in the Frame System Editor](img/demo_camera_edited_2.png)

Now that the Frame System is configured so that the spatial visualization for your robot matches its spatial orientation in real life, you can utilize [Motion Service](/services/motion) methods that take in reference frame information.

### Display Options

Click and drag on the **Frame System** visualization to view the display from different angles, and pinch to zoom in and out:

> {{<gif webm_src="img/frame_system_demo.webm" mp4_src="img/frame_system_demo.mp4" alt="The frame system visualization zooming in and out around the example robot with a camera, board, and wheeled base.">}}

Reviewer: this is very cool - for the future, open up the developer tools and remove the chat bubble when taking these screen recordings (let me know if you want me to show you how). Not worth redoing now, just keep it in mind

Author: Oh I didn't think of that thank you!

Click the video camera icon below and to the right of the **Frame System** button to switch beween the default **Perspective Camera** and the **Orthographic Camera** view:

{{< tabs name="Toggle Camera Views" >}}
{{% tab name="Perspective Camera" %}}

![Default Perspective Camera view shown in the Frame System Editor](img/demo_perspective.png)

{{% /tab %}}
{{% tab name="Orthographic Camera" %}}

![Non-default Orthographic Camera view shown in the Frame System Editor](img/demo_orthographic.png)

{{% /tab %}}
{{< /tabs >}}

### Bounding Geometries

> Add `Geometry` properties to the **Frame System** [configuration](#configuration) of a component to visualize bounding geometries by selecting a component and selecting a **Geometry** type.

Reviewer: Can you specify where I do that?

Author: done, it does look... chunky though! 😆

> By default, a **Geometry** is shown surrounding the origin point of a component as follows:

Reviewer: can you describe this? Maybe "as a cube around the component"

Author: it's not necessarily a box though!

Author: Not necessarily default as box but removed as follows

{{< tabs name="Visualize Adding Geometry Bounds" >}}
{{% tab name="Box" %}}

![Demo robot with default box bounds added to the wheeled base, shown in the Frame System Editor](img/demo_bound_box.png)

> You can adjust the **Size** and **Translation** of a **Geometry** to change these bounds.

Author: It doesn't sound too odd to me but happy to change it, easiest way I can think of to refer to the feature

> You can adjust the **Size** and **Translation** of a **Geometry** to change these bounds.

Author: It is a little weird though

For example:

Consider a robot configured with a [`jetson` board](/components/board), wired to a [`webcam` camera](/components/camera/webcam) and a [`wheeled` base](/components/base/wheeled) with two [motors](/components/motor) driving its wheels.

You have not yet added any reference frame configuration to these components.
If you go to the **config** tab of your robot and click on the **Frame System** sub-tab, they are all shown to be located on the default `world` origin point as follows:

  ![Example robot's default frame configuration shown in the Frame System Editor. All components are stuck on top of each other](img/demo_base_unedited.png)

You measure the distance on the floor from your base to your board and camera setup and find it to be 200 millimeters.
If you add this value to `"X"` in your base's reference frame `Translation` attribute, the Frame System readjusts to show the base's translation:

  ![Base translated 200mm forwards shown in the Frame System Editor](img/demo_base_edited.png)

Next, you measure the distance from your board to the camera mounted overhead and find it to be 50 millimeters.

> If you add this value to `"Z"` in your camera's reference frame `Translation` attribute, the Frame System readjusts to show the camera's translation:

Reviewer: Since this is all an example I find the `you` strange in all of this.

Author: Yeah it's a little weird.

 I don't know how to translate directions like `You have not yet added any reference frame configuration to these components.` without speaking to the user or using *we*

{{% tab name="Box" %}}

![Demo robot with default box bounds added to the wheeled base, shown in the Frame System Editor](img/demo_bound_box.png)

You can adjust the **Size** and **Translation** of a **Geometry** to change these bounds.
For example:

![Demo robot with translated box bounds added to the wheeled base, shown in the Frame System Editor](img/demo_bound_box_translation.png)

{{< /tab >}}
{{% tab name="Sphere" %}}
![Demo robot with default sphere bounds added to the wheeled base, shown in the Frame System Editor](img/demo_bound_sphere.png)

{{% /tab %}}
{{% tab name="Capsule" %}}

![Demo robot with default capsule bounds added to the wheeled base, shown in the Frame System Editor](img/demo_bound_capsule.png)

{{% /tab %}}
{{< /tabs >}}

## How the Frame System Works

`viam-server` builds a tree of reference frames for your robot with the `world` as the root node and regenerates this tree following reconfiguration.

Access a [topologically-sorted list](https://en.wikipedia.org/wiki/Topological_sorting) of the generated reference frames in the robot's logs at `--debug` level:

![an example of a logged frame system](img/frame_sys_log_example.png)

> Consider the example of nested reference frame configuration where a [two dynamic components are attached](/services/frame-system/nested-frame-config): a robotic arm, `A`, attaches to a gantry, `G`, which in turn is fixed in place at a point on the `World` of a table.

Author: ```suggestion
Consider the example of nested reference frame configuration where [two dynamic components are attached](/services/frame-system/nested-frame-config): a robotic arm, `A`, attaches to a gantry, `G`, which in turn is fixed in place at a point on the `World` of a table.
```
